// sagas/index.js
import {call, put, takeEvery} from "redux-saga/effects";
import {fetchPhonesApi, fetchPhoneByIdApi, fetchCategoriesApi, editPhoneByIdApi, addPhoneApi, deletePhoneApi} from "api";
import {requestPhonesError, requestPhonesSuccess, fetchPhoneByIdError, fetchPhoneByIdSuccess,fetchCategoriesSuccess,fetchCategoriesError, editPhoneIdError, editPhoneIdSuccess, addPhoneSuccess, addPhoneError, deletePhoneSuccess, deletePhoneError} from "actions";
import * as actionType from "../actionTypes";

export default function* watchFetchData() {
    yield takeEvery(actionType.FETCH_PHONES_START, fetchPhones);
    yield takeEvery(actionType.FETCH_PHONE_BY_ID_START, fetchPhoneById);
    yield takeEvery(actionType.FETCH_CATEGORIES_START, fetchCategories);
    yield takeEvery(actionType.GET_EDIT_PHONE_BY_ID,editPhoneId)
    yield takeEvery(actionType.ADD_PHONE_START, addPhone)
    yield takeEvery(actionType.DELETE_PHONE_START, deletePhone)
}

function* fetchPhones() {
    try {
        const phones = yield call(fetchPhonesApi)
        yield put(requestPhonesSuccess(phones))
    } catch (error) {
        yield put(requestPhonesError())
    }
}

function* fetchPhoneById(action) {
    try {
        const phone = yield call(fetchPhoneByIdApi, action.id)
        yield put(fetchPhoneByIdSuccess(phone))
    } catch (error) {
        yield put(fetchPhoneByIdError())
    }
}

function* fetchCategories() {
    try {
        const categories = yield call(fetchCategoriesApi)
        yield put(fetchCategoriesSuccess(categories))
    } catch (error) {
        yield put(fetchCategoriesError())
    }
}

function* editPhoneId(action) {
    try {
        yield call(editPhoneByIdApi, action.id , action.data)
        yield put(editPhoneIdSuccess(action.id , action.data))
    } catch (error) {
        yield put(editPhoneIdError())
    }
}

function* addPhone(action) {
    try {
        yield call(addPhoneApi, action.data , action.id)
        yield put(addPhoneSuccess(action.data , action.id))
    } catch (error) {
        yield put(addPhoneError())
    }
}

function* deletePhone(action) {
    try {
        yield call(deletePhoneApi,  action.id)
        yield put(deletePhoneSuccess(action.id))
    } catch (error) {
        yield put(deletePhoneError())
    }
}
